src/toxcore/tcp/packet.rs

```diff
@@ -0,0 +1,514 @@
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*! Top-level TCP Packets according
+    to https://zetok.github.io/tox-spec/#encrypted-payload-types
+*/
+
+use toxcore::crypto_core::*;
+use toxcore::tcp::binary_io::*;
+use nom::*;
+use cookie_factory::*;
+
+/** Top-level TCP packet.
+
+    According to https://zetok.github.io/tox-spec/#encrypted-payload-types.
+*/
+#[derive(Debug, PartialEq, Clone)]
+pub enum Packet {
+    /// [`RouteRequest`](./struct.RouteRequest.html) structure.
+    RouteRequest(RouteRequest),
+    /// [`RouteResponse`](./struct.RouteResponse.html) structure.
+    RouteResponse(RouteResponse),
+    /// [`ConnectNotification`](./struct.ConnectNotification.html) structure.
+    ConnectNotification(ConnectNotification),
+    /// [`DisconnectNotification`](./struct.DisconnectNotification.html) structure.
+    DisconnectNotification(DisconnectNotification),
+    /// [`PingRequest`](./struct.PingRequest.html) structure.
+    PingRequest(PingRequest),
+    /// [`PongResponse`](./struct.PongResponse.html) structure.
+    PongResponse(PongResponse),
+    /// [`OobSend`](./struct.OobSend.html) structure.
+    OobSend(OobSend),
+    /// [`OobReceive`](./struct.OobReceive.html) structure.
+    OobReceive(OobReceive),
+    /// TODO
+    //OnionDataRequest,
+    /// TODO
+    //OnionDataResponse,
+    /// [`Data`](./struct.Data.html) structure.
+    Data(Data)
+}
+
+/// A serialized Packet should be not longer than 2032 bytes
+pub const MAX_TCP_PACKET_SIZE: usize = 2032;
+
+impl FromBytes for Packet {
+    named!(from_bytes<Packet>, alt!(
+        map!(RouteRequest::from_bytes, Packet::RouteRequest) |
+        map!(RouteResponse::from_bytes, Packet::RouteResponse) |
+        map!(ConnectNotification::from_bytes, Packet::ConnectNotification) |
+        map!(DisconnectNotification::from_bytes, Packet::DisconnectNotification) |
+        map!(PingRequest::from_bytes, Packet::PingRequest) |
+        map!(PongResponse::from_bytes, Packet::PongResponse) |
+        map!(OobSend::from_bytes, Packet::OobSend) |
+        map!(OobReceive::from_bytes, Packet::OobReceive) |
+        map!(Data::from_bytes, Packet::Data)
+    ));
+}
+
+impl ToBytes for Packet {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        match self {
```

Comment:
Don't clippy warn that you can use *self?

Reply:
I did not run it. How do I do that?

Reply:
You can run cargo build --features "clippy" with nightly rust. Or if you update clippy it might work on travis so you can see warnings there.